experimental/packages/opentelemetry-instrumentation/src/instrumentation.ts

```diff
@@ -173,4 +175,33 @@ export abstract class InstrumentationAbstract implements Instrumentation {
     | InstrumentationModuleDefinition
     | InstrumentationModuleDefinition[]
     | void;
+
+  /**
+   * Execute instrumentation event hook, if configured, and log any errors.
+   * Any semantics of the event type and eventInfo are defined by the specific instrumentation.
+   * @param hookHandler The optional hook handler which the user has configured via instrumentation config\
+   * @param eventName The name of the event being triggered, for logging purposes
+   * @param span The span to which the hook should be applied
+   * @param eventInfo The event info to be passed to the hook
+   */
+  protected _runInstrumentationEventHook<EventInfoType>(
```

Comment:
Hmm, I wonder if we can make this whole protected method actually a stand-alone function.

As a protected method it's actually part of the public interface anyway, so adding it as a stand-alone function does turn out to be the same. I'm worried that, as more signals become available, we add a lot of util functions to the instrumentation base that will make it blow up in size over time.

Adding things to abstract classes is also difficult once stable, as in typescript consumers can use the type like:

```ts
// uses instrumentation 1.1 (new protected method in InstrumentationBase)
import {InstrumentationBase} from '@opentelemetry/instrumentation';
// uses instrumentation 1.0 (missing protected method from 1.1, uses pinned for some reason 1.0 as dependency)
import {MyInstrumentation} from '@opentelemetry/my-cool-instrumentation';

// this assignment will not compile as InstrumentationBase (1.1) is defining more protected properties than InstrumentationBase (1.0) used by MyInstrumentation
const myInstrumentation: InstrumentationBase = new MyInstrumentation();

```

We've seen this happen with the (non-abstract) Resource class #3676

I feel putting this here now may spark lots of requests later to add more causing us to run into a similar problem (I'll be less impactful than the Resource one, as I don't assume people use it like I described above a lot, but it can still happen and can be breaking).

It's actually something that happens occasionally with `MetricReader` too, especially with `@opentelemetry/sdk-node` where `@opentelemetry/sdk-metrics` is pinned.

Reply:
I would love to have it as util, but since it's using `this._diag` I need access to `this` which felt like a class member function.

Having it as util would mean passing also the diag logger as an argument to the function, which felt strange to me.

WDYT?

Reply:
> Adding things to abstract classes is also difficult once stable, as in typescript consumers can use the type like:
> 
> ```ts
> // uses instrumentation 1.1 (new protected method in InstrumentationBase)
> import {InstrumentationBase} from '@opentelemetry/instrumentation';
> // uses instrumentation 1.0 (missing protected method from 1.1, uses pinned for some reason 1.0 as dependency)
> import {MyInstrumentation} from '@opentelemetry/my-cool-instrumentation';
> 
> // this assignment will not compile as InstrumentationBase (1.1) is defining more protected properties than InstrumentationBase (1.0) used by MyInstrumentation
> const myInstrumentation: InstrumentationBase = new MyInstrumentation();
> ```

For the code snippet above, I guess it would make more sense to use the `Instrumentation` interface and not type something as `InstrumentationBase` directly, although no one will stop users from doing this. 

Do you think once stable we should avoid adding new methods to `InstrumentationAbstract`, or are we still allowed to do that but should try not to if possible?

Reply:
> I would love to have it as util, but since it's using this._diag I need access to this which felt like a class member function.

Hmm, I see, let's leave it like this then and re-evaluate later.

> Do you think once stable we should avoid adding new methods to InstrumentationAbstract, or are we still allowed to do that but should try not to if possible?

Hard to say. I think we should at least try not to do it. I've tried to find resources on this topic (specifically related to typescript libraries) but I was unsuccessful in finding a concrete answer to the question (which I've had now for quite a while). Most of the resources online and in books are specifically related to TypeScript in apps which is unfortunately not as helpful for my specific questions as a library author.

From what I've seen users expect that these breakages should not occur outside of major versions and, intuitively, I agree. For the `MetricReader` we've done it before simply because that's the only way we could make any changes at all. Impact on end-users is fairly minimal.

For TypeScript, it is extremely easy to break users when exporting classes, this is further complicated by having situations that allow for multiple versions of the same package, and the fact that private class members are part of the public interface, too. Adding anything is a nightmare when being strict enough.

What that'd mean for us is that:
- we should not export classes as a public API
  - we can only use functions (factory functions) and interfaces, which we may extend with optional properties/parameters
- we should also not export abstract classes, but work interfaces and composition (injecting a helper through an interface method) where possible

I'm not sure if that's right or the the way to go though. Let's leave this PR as-is and discuss this another time. It's likely a larger change that will affect all components and needs to be carefully considered.

Reply:
Thanks for sharing your thoughts and describing the problem so clearly. I agree, it's best to leave this PR as-is and reevaluate later 🙏

> What that'd mean for us is that:
> 
> * we should not export classes as a public API
>   
>   * we can only use functions (factory functions) and interfaces, which we may extend with optional properties/parameters
> * we should also not export abstract classes, but work interfaces and composition (injecting a helper through an interface method) where possible

Trying to think how these considerations can be applied in the InstrumentationBase case. would love to discuss this later on